schema/query/predicate.go

```diff
@@ -216,12 +222,18 @@ func (e NotIn) String() string {
 	for _, v := range e.Values {
 		s = append(s, valueString(v))
 	}
-	return quoteField(e.Field) + ": {" + opNotIn + ": [" + strings.Join(s, ", ") + "]}"
+	return quoteField(e.Field) + ": {" + schema.OpNotIn + ": [" + strings.Join(s, ", ") + "]}"
 }
 
 // Validate implements Expression interface.
 func (e NotIn) Validate(validator schema.Validator) error {
-	return validateValues(e.Field, e.Values, validator)
+	return validateValues(e.Field, e.Values, schema.OpNotIn, validator)
+}
+
+// Fix TODO
+func (e NotIn) Fix(validator schema.Validator) (Expression, error) {
+	newValues, err := validateValuesOperator(e.Field, e.Values, schema.OpNotIn, validator)
+	return NotIn{Field: e.Field, Values: newValues}, err
```

Comment:
It looks to me like you should set the "newValues" during the initial ParsePredicate -- is there a reason that won't work?

Reply:
Right, I see the problem now. I will try to find some time and look at the code a bit deeper, and then get back to you. I suppose it _could_ maybe be done as part of Validate if it's changed to use pointer receivers all over the place...

I will try to find some time over the weekend.

Reply:
Thanks for your work!

Reply:
I started looking at where `ParsePredicate` is being called, and that gives away waht's availabe at these locations in the code -- e.g. is the schema validator available?

These are the non-test results I found

https://github.com/rs/rest-layer/blob/48c11e50091e8c988780853a3aaa8af25cf98d42/schema/query/query.go#L69

https://github.com/rs/rest-layer/blob/9197c24a34866e9717ca55c4364df76c362d95dc/rest/routing.go#L261

https://github.com/rs/rest-layer/blob/48c11e50091e8c988780853a3aaa8af25cf98d42/graphql/query.go#L114

https://github.com/rs/rest-layer/blob/48c11e50091e8c988780853a3aaa8af25cf98d42/graphql/query.go#L125


The first case is query.New, which is a bit of a special case, and I can't find it used outside of tests.

The next three usages all have the `schema.Validator` interface available, and in fact, the `.Validate` method is called on each instance here.

I am thinking all of the following are possible solutions that would work, and we can discuss the design:

1. Validate is changed to use pointer receivers, and modifes the parent predicate/expression.
2. Validate is set to return a new predicate -- it can optionally be ignored when it's not needed.
3. A second inializer, ParsePredicateForValidator(schema.Validator) is introduced.

In the case of design, I feel 2 is the best fit for rest-layer for the following reasons:
- The mutation is explicit.
- The interface is kept similar to the interface of schema.Validate, that also returns validated type and an error.
- All the other query components (ordering and projection) calls .Validate -- continuing to call .Validate is the least amont of change that would allow for a, at least mostly, consistent API.

@rs -- If you have a view on the design of this, it would be great to know.

The problem we want to solve, is more or less the one described in #170. I believe @Dragomir-Ivanov is particular keen on solving it for the case of `$lt` and `$gt` for time.Time, and potentially a custom `Deciaml` type. Solving it in a generic way would be nice, so that it's compatible with third-party FieldValidators and database storage types.

Reply:
I intentionally skipped the MustParsePredicate.

It's all non-tests inside _rest-layer_ where it's called. `MustParsePredicate` would probablly be called in end-consumer applications, and would need to stay either way, as it's recommended for use in `schema` definitions etc.

Reply:
Yes, but how we are going to proceed in this case. There is no `schema.Validator` declaration in this context? `ParsePredicate` may accept it as a pointer value, so we can pass `nil` in case whether it is not accessible. However I feel that we need always have the schema validator where we are making a query.

Reply:
> ParsePredicate may accept it as a pointer value, so we can pass nil in case whether it is not accessible. However I feel that we need always have the schema validator where we are making a query.

It's mostly available, but there seem to be two exceptions.

1. When you defined a Field Dependency - This is the main use-case for `MustParsePredicate`, the schema isn't ready yet, so no schema.Validator can be passed in at parse-time.
2. The internal code for evaluating a projection on a connected field (the last `ParsePredicate` code reference I previously overlooked),  don't have access to the related schema today, and it does not look like it even calls validate on the parsed predicate before using it.

Case 1, from what I can see, get's validated later:
- https://github.com/rs/rest-layer/blob/48c11e50091e8c988780853a3aaa8af25cf98d42/schema/dependency.go#L23

Case 2 doesn't seam to be validated at all today (unless I somehow missed it), which seams like a bug.

Reply:
Check that it's the right one though... you want the schema.Validator of the connected schema, not the schema you are _at_.

Connections are used to _embed_ other resources.

Reply:
https://github.com/rs/rest-layer#embedding